nidx/nidx_relation/src/graph_collector.rs

```diff
@@ -0,0 +1,229 @@
+// along with this program. If not, see <http://www.gnu.org/licenses/>.
+//
+
+use std::collections::{HashMap, HashSet};
+
+use tantivy::{
+    DocId, Score, SegmentOrdinal, SegmentReader, TantivyDocument,
+    collector::{Collector, SegmentCollector},
+    store::StoreReader,
+};
+use tracing::warn;
+
+use crate::io_maps;
+
+// XXX: this is a completely arbitrary number. Feel free to change for a more adequate value
+const STORE_READER_CACHED_BLOCKS: usize = 10;
+
+pub type NodeId = (String, i32, String);
+pub type RelationId = (String, u64);
+
+#[derive(Clone, Copy)]
+pub enum NodeSelector {
+    AnyNode,
+    SourceNodes,
+    DestinationNodes,
+}
+
+pub struct TopUniqueNodeCollector {
+    limit: usize,
+    selector: NodeSelector,
+    schema: crate::schema::Schema,
+}
+
+pub struct TopUniqueRelationCollector {
+    limit: usize,
+    schema: crate::schema::Schema,
+}
+
+pub struct TopUniqueNodeSegmentCollector {
+    limit: usize,
+    selector: NodeSelector,
+    unique: HashSet<NodeId>,
+    schema: crate::schema::Schema,
+    store_reader: StoreReader,
+}
+
+pub struct TopUniqueRelationSegmentCollector {
+    limit: usize,
+    unique: HashMap<RelationId, TantivyDocument>,
+    schema: crate::schema::Schema,
+    store_reader: StoreReader,
+}
+
+impl TopUniqueNodeCollector {
+    pub fn new(selector: NodeSelector, limit: usize) -> Self {
+        Self {
+            limit,
+            selector,
+            schema: crate::schema::Schema::new(),
+        }
+    }
+}
+
+impl TopUniqueRelationCollector {
+    pub fn new(limit: usize) -> Self {
+        Self {
+            limit,
+            schema: crate::schema::Schema::new(),
+        }
+    }
+}
+
+impl Collector for TopUniqueNodeCollector {
+    type Fruit = HashSet<NodeId>;
+    type Child = TopUniqueNodeSegmentCollector;
+
+    fn requires_scoring(&self) -> bool {
+        true
+    }
+
+    fn for_segment(&self, _segment_local_id: SegmentOrdinal, segment: &SegmentReader) -> tantivy::Result<Self::Child> {
+        Ok(TopUniqueNodeSegmentCollector {
+            limit: self.limit,
+            selector: self.selector,
+            unique: HashSet::new(),
+            store_reader: segment.get_store_reader(STORE_READER_CACHED_BLOCKS)?,
+            schema: self.schema.clone(),
+        })
+    }
+
+    fn merge_fruits(
+        &self,
+        segment_fruits: Vec<<Self::Child as SegmentCollector>::Fruit>,
+    ) -> tantivy::Result<Self::Fruit> {
+        let mut unique = HashSet::new();
+        let mut fruits = segment_fruits.into_iter().flatten();
+        let mut fruit = fruits.next();
+
+        while fruit.is_some() && unique.len() < self.limit {
+            unique.insert(fruit.unwrap());
+            fruit = fruits.next();
+        }
+        Ok(unique)
+    }
+}
+
+impl Collector for TopUniqueRelationCollector {
+    type Fruit = Vec<TantivyDocument>;
+    type Child = TopUniqueRelationSegmentCollector;
+
+    fn requires_scoring(&self) -> bool {
+        true
```

Comment:
Given that you are not using the score at all, I'd suggest setting this to false (same for the other collector). It avoids calculating scores just to be discarded and it's a bit faster.

Reply:
In this matter, I think to be correct we should consider scores and only return the top K from best to worse, as fuzzy could cause weird behavior.